---
title: "SCopeLoomR Tutorial - Create a .loom file and extract data from a .loom"
package: r pkg_ver('SCopeLoomR')
output:
  html_notebook:
    toc: yes
  html_document:
    keep_md: true
    df_print: paged
    toc: yes
  BiocStyle::html_document:
    number_sections: no
  pdf_document:
    toc: yes
vignette: |
  %\VignetteIndexEntry{SCopeLoomR} 
  %\VignetteEncoding{UTF-8} 
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(SCopeLoomR)
library(SingleCellExperiment)
```

# Introduction

The .loom format is a file format designed to store very large omics datasets. It has been created is maintained by the [Linnarsson lab](http://linnarssonlab.org/). `SCopeLoomR` is an R package that helps you to create easily .loom files that follow the data [naming conventions](http://linnarssonlab.org/loompy/conventions/index.html) from loompy. 

`SCopeLoomR` was created along the development of SCope to help to store different data inside the .loom e.g.: regulon, regulon activities from SCENIC;  . Therefore, it 

# Installation
For example:
```{r install, eval=FALSE}
devtools::install_github("aertslab/SCopeLoomR")
```

# Creating a loom object

## Load Data

(minimum information required to create the object)

```{r loadData}
library(SingleCellExperiment)
library(SCopeLoomR)
data(sce) # e.g. a SingleCellExperiment 

# DGEM (Digital gene expression matrix)
dgem <- counts(sce)
  dim(dgem)
  class(dgem)
  head(colnames(dgem))  #should contain the Cell ID/name

# Known cell information/annotation  
cell.info <- colData(sce)
cell.info$nGene <- colSums(dgem>0)
  head(cell.info)

# Default embedding (e.g. t-SNE or PCA coordinates)
data(tSNE_expr)
default.tsne <- tSNE_expr
default.tne.name <- "t-SNE on full expression matrix"
head(default.tsne)
```

## Create loom file 

```{r createLoom}
library(SCopeLoomR)

### Create the minimal loom file
file.name <- "example.loom"
loom <- build_loom(file.name=file.name,
                   dgem=dgem,
                   title="Fake expression dataset for examples",
                   genome="Mouse", # Just for user information, not used internally
                   default.embedding=default.tsne,
                   default.embedding.name=default.tne.name)
```

## Add hierarchy

You can organize/group your .loom files in SCope by specifiying differents grouping levels. The current .loom file will be put in `Mouse -> Toy Datasets` branch of the SCope loom tree. 

```{r}
add_hierarchy(loom = loom, hierarchy = create_hierarchy(level.1.name = "Mouse", level.2.name = "Toy Datasets", level.3.name = ""))
```

## Add annotations/metrics

You can add different annotations and/or metrics that you will be able to query in SCope.

```{r}
# Add annotation
add_col_attr(loom=loom, key = "Cell type", value=cell.info$cellType, as.annotation=T)
# Add metric
add_col_attr(loom=loom, key = "nGene", value=cell.info$nGene, as.metric = T)
```

## Add SCENIC results (Optional)

*(Not run in this example)*

```{r addSCENIC, eval=FALSE}
scenic.dir <- "./" # Modify if needed

# Regulons AUC matrix
library(AUCell)
load(paste0(scenic.dir, "int/3.2_regulonAUC.RData"), verbose=TRUE)
regulons.AUC <- updateAucellResults(regulonAUC)
add_scenic_regulons_auc_matrix(loom=loom, regulons.AUC=getAUC(regulons.AUC))

# Regulons (gene list), regulon thresholds (optional) and regulon motifs (optional)
load(paste0(scenic.dir, "int/3.0_regulons_forAUCell.RData"), verbose=TRUE)
regulon.enrichment.table<-read.table(file = "output/Step2_MotifEnrichment.tsv", header = T, sep = "\t", quote = '', stringsAsFactors = F)
add_scenic_regulons(loom=loom
                    , regulons=regulons
                    , regulon.threshold.assignments=...$threshold.assignment # Optional
                    , regulon.enrichment.table = regulon.enrichment.table # Optional) 

# Alternative t-SNE
load(paste0(scenic.dir, "int/3.3_tsneRegulonAUC_PCA_50pc50perpl.RData"))
add_embedding(loom=loom, embedding=tSNE$Y, name="SCENIC (t-SNE on AUC)")
```

## Add Seurat results (Optional)

*(Not run in this example. You would need to use your own file names)*

```{r addSEURAT, eval=FALSE}
seurat.dir <- "Seurat_output/"
seurat.tsne <- readRDS(file = paste0(seurat.dir, "seurat_tsne.rds.gz"))
seurat <- readRDS(paste0(seurat.dir, "seuratObject.rds.gz"))

# Add extra embeddings
add_embedding(loom=loom, embedding=seurat.tsne, name="Seurat 82PC, 30perp"))
```

### Add clusterings

```{r}
add_seurat_clustering(loom = loom
                      , seurat = seurat)
```

### Add clustering(s) along with annotation for a given resolution (default one if set)
```{r}
seurat.annotation<-read.table(file = paste0(seuratDir, "Res2_Clusters.tsv", header=T, quote = '', sep = "\t", stringsAsFactors=F))
add_seurat_clustering(loom = loom
                      , seurat = seurat
                      , default.clustering.resolution = "res.2"
                      , annotation = seurat.annotation
                      , annotation.cluster.id.cn = "res.2" 
                      , annotation.cluster.description.cn = "Annotation")
```

### Add clustering(s) along with cluster markers 
```{r}
seurat.resolutions <- get_seurat_clustering_resolutions(seurat = seurat)
seurat.markers.file.path.list<-do.call(what = "c", lapply(X=seurat.resolutions, FUN=function(resolution) {
  l<-list()
  l[paste0("res.",resolution)]<-paste0(seuratDir, "res.",resolution,"/Seurat_PC82_res",resolution,"_bimod_Cluster_Markers.rds.gz")
  return (l)
}))
add_seurat_clustering(loom = loom
                      , seurat = seurat
                      , seurat.markers.file.path.list = seurat.markers.file.path.list)
```

### Add clustering(s), along with cluster markers and metrics (e.g.: logFC, p-value, ...)

```{r}
add_seurat_clustering(loom = loom
                      , seurat = seurat
                      , seurat.markers.file.path.list = seurat.markers.file.path.list
                      , seurat.marker.metric.accessors = c("avg_logFC", "p_val_adj")
                      , seurat.marker.metric.names = c("Avg. logFC", "Adj. p-value")
                      , seurat.marker.metric.description = c("Average log fold change from Wilcox differential test (cfr. Seurat)"
                                                             , "Adjusted p-value using Bonferroni correction based on the total number of genes in the dataset (cfr. Seurat)"))
```

## Add single-cell trajectory data

### Monocle
Add the trajectory data from a CellDataSet object generated by monocle.

```{r}
# Add monocle embedding and trajectory
S_matrix <- reducedDimS(cds)
monocle.embedding<-data.frame(t(S_matrix[c(1, 2), ]))
add_embedding(loom = loom, embedding = monocle.embedding, name = "Monocle (DDRTree)", trajectory = create_trajectory_from_monocle(cds = cds))
```


## Finalize (save)
```{r finalize}
finalize(loom=loom)
```

# Extract data from a loom object

The .loom file related to this part of the tutorial can be downloaded at http://scope.aertslab.org in the left panel under `Drosophila` > `Brain`.

## Get the gene expression matrix

```{r}
loom<-open_loom(loom = "Aerts_Fly_AdultBrain_Filtered_57k.loom")
dgem<-get_dgem(loom = loom)
```

## Get the gene expression matrix of a given cluster

```{r}
loom<-open_loom(loom = "Aerts_Fly_AdultBrain_Filtered_57k.loom")
cluster.10.dgem<-get_cluster_dgem_by_name(loom = loom, cluster.name = "Astrocyte-like - Cluster 10")
```




